pkg/operator/operator.go

```diff
@@ -743,3 +786,337 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
+			return err
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *Operator) rebalanceWorkload(ctx context.Context, pod *v1.Pod, pvc *v1.PersistentVolumeClaim) (bool, error) {
+	node, err := o.client.GetNode(ctx, pod.Spec.NodeName)
+	if err != nil {
+		return false, err
+	}
+
+	// Check that the node wasn't already cordoned to prevent tempering someone else cordon.
+	alreadyCordoned := node.Spec.Unschedulable
+	if !alreadyCordoned {
+		klog.V(2).Infof("Cordoning node %s.", node.Name)
+		err = drain.RunCordonOrUncordon(o.drainer, node, true)
+		if err != nil {
+			return false, err
+		}
+		// Set annotation so that we know CMO was at the origin of the cordon
+		klog.V(4).Infof("Adding annotation %s to node %s since it was cordoned by CMO.", cordonAnnotation, node.Name)
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			node, err := o.client.GetNode(ctx, node.Name)
+			if err != nil {
+				return err
+			}
+			if node.Annotations == nil {
+				node.Annotations = make(map[string]string)
+			}
+			node.Annotations[cordonAnnotation] = fmt.Sprintf("node marked as unschedulable by cluster-monitoring-operator to reschedule %s/%s on another node", pod.Namespace, pod.Name)
+			_, err = o.client.UpdateNode(ctx, node)
+			return err
+		})
+		if err != nil {
+			return false, err
+		}
+	}
+
+	pv, err := o.client.GetPersistentVolume(ctx, pvc.Spec.VolumeName)
+	if err != nil {
+		return false, err
+	}
+	if pv.Labels != nil {
+		// Do not delete the PVC if the storage provider hasn't set the topology.kubernetes.io/zone label on the PV.
+		// In most cases, when the PV isn't zonal, pods can access it from a node in a different AZ so we don't need to delete it.
+		if _, ok := pv.Labels[zonalTopologyAnnotation]; ok {
+			klog.V(2).Infof("Deleting PersistentVolumeClaim %s/%s.", pvc.Namespace, pvc.Name)
+			err = o.client.DeletePersistentVolumeClaim(ctx, pvc)
+			if err != nil {
+				return false, err
+			}
+		}
+	}
+
+	klog.V(2).Infof("Deleting pod %s/%s.", pod.Namespace, pod.Name)
+	err = o.client.DeletePod(ctx, pod)
+	if err != nil {
+		return false, err
+	}
+
+	if !alreadyCordoned {
+		klog.V(2).Infof("Uncordoning node %s.", node.Name)
+		err = drain.RunCordonOrUncordon(o.drainer, node, false)
+		if err != nil {
+			return false, err
+		}
+		klog.V(4).Infof("Removing annotation %s from node %s since it was uncordoned.", cordonAnnotation, node.Name)
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			node, err := o.client.GetNode(ctx, node.Name)
+			if err != nil {
+				return err
+			}
+			delete(node.Annotations, cordonAnnotation)
+			_, err = o.client.UpdateNode(ctx, node)
+			return err
+		})
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+func (o *Operator) rebalanceWorkloads(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
```

Comment:
Should we keep only pods in `Running` state?

Reply:
I don't think there is a real gain in doing that, at worst we might even end up skipping pods that should have been rescheduled.